### PR TITLE
COSE Encrypt: Redesign of Recipient Structure

### PR DESCRIPTION
**Pull request by hannestschofenig**
This PR contains COSE encryption functionality and offers key distribution via HPKE and AES Key Wrap. 

Note: A description of HPKE in COSE can be found in https://datatracker.ietf.org/doc/html/draft-ietf-cose-hpke and the used HPKE library is available at https://github.com/hannestschofenig/mbedtls/tree/hpke.

Important: This is work in  progress. Expect changes to the API and to the internals of the implementation. 

Change-Id: I190f58a82eb4a883e04f1a5cf17cd40bd27ea9ff

### CODE REVIEW

File: src/t_cose_aes_kw.c [hunk 1]
```diff
@@ -0,0 +1,103 @@
+/**
+ * \file t_cose_aes_kw.c
+ *
```
Comment (laurencelundblade) [inline src/t_cose_aes_kw.c:3]: Rename to t_cose_recipient_enc_kw.{ch}

File: inc/t_cose/t_cose_hpke.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*
+ * t_cose_hpke.h
```
Comment (laurencelundblade) [inline inc/t_cose/t_cose_hpke.h:2]: Rename to t_cose_recipient_enc_hpke.{ch}

**Review by laurencelundblade — CHANGES_REQUESTED** [review]
The requested changes here are focused just on what is needed to merge this PR. They are mostly about dependency and common files like Makefiles that are subject to a lot of churn from other inbound PRs.

Once encrypt is merged, then there will be other things to work on in the design and implementation, but they can be handled in smaller and easier chunks.

File: crypto_adapters/t_cose_psa_crypto.c [hunk 3]
```diff
@@ -36,7 +38,8 @@
 
 #include "t_cose_crypto.h"  /* The interface this implements */
 #include <psa/crypto.h>     /* PSA Crypto Interface to mbed crypto or such */
-
+#include <mbedtls/aes.h>
+#include <mbedtls/nist_kw.h>
```
[the remark this answers is not in the document]
Comment (hannestschofenig) [reply crypto_adapters/t_cose_psa_crypto.c:42]: Unfortunately, there are no PSA APIs for AES-KW and for HPKE. However, the implementation of AES-KW and HPKE uses PSA APIs internally.

**Comment by hannestschofenig** [comment]
To address the comments by Laurence and the changes in the t_cose codebase since the submission of this PR I created a new PR, which can be found here: #115

**Comment by laurencelundblade** [comment]
replaced by https://github.com/laurencelundblade/t_cose/pull/116

**Pull request closed without merging**